raft/src/main/java/org/apache/kafka/raft/Batch.java

```diff
@@ -33,21 +33,24 @@
     private final int sizeInBytes;
     private final long lastOffset;
     private final List<T> records;
+    private final List<ControlRecord> controlRecords;
```

Comment:
I thought that batches either contained all control records or none. Is that not the case? If that is so, should we add some invariants here to enforce that?

Reply:
Yes. That is correct. A record batch has 1 or more control records; or 1 or more data records.  We enforce that in two ways. 1) the ctr for this type is private and 2) the static methods `data` and `control` check that the passed list of records have one or more records.